docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,154 @@
+	unsavedFastNodeAdditions map[string]*fastnode.Node
+	unsavedFastNodeRemovals  map[string]interface{}   
+	ndb                      *nodeDB
+	skipFastStorageUpgrade   bool 
+
+	mtx sync.Mutex
+}
+```
+
+We will assign the `nodeKey` when saving the current version in `SaveVersion`. It will reduce unnecessary checks in CRUD operations of the tree and keep sorted the order of insertion in the LSM tree.
+
+### Migration
+
+We can migrate nodes one by one by iterating the version.
+
+- Iterate the version in order, and get the root node for the specific version.
+- Iterate the tree based on the root and pick only nodes the node version is the same as the given version.
+- Store them using the new node key format.
+
+### Pruning
+
+We assume keeping only the range versions of `fromVersion` to `toVersion`. Refer to [this issue](https://github.com/cosmos/cosmos-sdk/issues/12989).
+
+Here we are introducing a new way how to get orphaned nodes which remove in the `n+1`th version updates.
+
+- Traverse the tree in-order way based on the root of `n+1`th version.
+- If we visit the lower version node, pick the node and don't visit further deeply. Pay attention to the order of these nodes.
+- Traverse the tree in-order way based on the root of `n`th version.
+- Iterate the tree until meet the first node among the above nodes(stack) and delete all visited nodes so far from `n`th tree.
+- Pop the first node from the stack and iterate again.
+
+If we assume `1 to (n-1)` versions already been removed, when we want to remove the `n`th version, we can just remove the above orphaned nodes.
+
+### Rollback
+
+When we want to rollback to the specific version `n`
+
+- Iterate the version from `n+1`.
+- Traverse key-value through `traversePrefix` with `prefix=bigendian(version)`.
+- Remove all iterated nodes.
+
+## Consequences
+
+### Positive
+
+Using the version and the path, we take advantage of data locality in the LSM tree. Since we commit the sorted data, it can reduce compactions and makes it easy to find the key. Also, it can reduce the key and node size in the storage.
+
+```
+# node body
+
+add `hash`:					    +32 byte
+add `leftNodeKey`, `rightNodeKey`:	max 8 + 8 = +16 byte
+remove `leftHash`, `rightHash`:			    -64 byte
+remove `version`: 			max	    -8	byte
+------------------------------------------------------------
+				total save	     24	byte
+
+# node key
+
+remove `hash`:				-32 byte
+add `version|path`:			+16 byte
+------------------------------------
+			total save 	 16 byte
+```
+
+Removing orphans also provides performance improvements including memory and storage saving.
+
+### Negative
+
+The `Update` operation will require extra DB access because we need to take children to calculate the hash of updated nodes.
+It doesn't require more access in other cases including `Set`, `Remove`, and `Proof`.
+
+It is impossible to remove the individual version. The new design requires more restrict pruning strategies.
```

Comment:
can you elaborate on this?

Reply:
I think it is enough, it will remove `orphans` and it will be impossible to remove the intermediate version from storage.
And `pruning` part explains in more detail the new methods.

Reply:
Even if removing orphaned, it's certainly possible to delete versions in between.

Reply:
lets edit to include this.